### PR TITLE
Flint univariate series expansion

### PR DESCRIPTION
**Pull request by rwst**
Flint polys cannot do `cot()`, `csc()`, and `lambert_w()` because they cannot handle Laurent polys.

### CODE REVIEW

**Comment by certik** [comment]
Great job. This looks good to me at first sight.

File: symengine/series_flint.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        hash = hash * 101  +  *s++;
+    }
+    return hash;
+}
+
+std::size_t URatPSeriesFlint::__hash__() const {
+    std::size_t seed = URATPSERIESPIRANHA;
+    hash_combine(seed, var_);
+    hash_combine(seed, degree_);
+    hash_combine(seed, str_hash(p_.to_string().c_str(), seed));
```
Comment (certik) [inline symengine/series_flint.cpp]: So in https://github.com/symengine/symengine/blob/3f0c0947ae09bac9545334a6903d9d72c7043a52/symengine/symbol.cpp#L14 we use:
```c++
std::hash<std::string> str_hash;
hash_combine(seed, str_hash(p_.to_string()));
```
is there a reason to use a different hash here?

**Comment by rwst** [comment]
Um, I get only 1370ms vs 1600ms (URatP-Piranha), not even 20% speedup with the `sincos-1000`. Callgrind will show.

File: CMakeLists.txt [hunk 2]
```diff
@@ -150,6 +155,19 @@ set(WITH_MPC no
...
+    include_directories(${FLINT_INCLUDE_DIRS})
+    set(LIBS ${LIBS} ${FLINT_TARGETS})
+    set(HAVE_SYMENGINE_FLINT True)
+    set(PKGS ${PKGS} "FLINT")
+
+    # Workaround for https://github.com/fredrik-johansson/arb/issues/24
+    include_directories(${FLINT_INCLUDE_DIRS}/flint)
+
+    set(WITH_MPFR yes)
+endif()
```
Comment (isuruf) [inline CMakeLists.txt]: If this block is after `WITH_ARB` then there's no need of code duplication.

File: symengine/series_flint.cpp (hunk 1 above)
Comment (rwst) [inline symengine/series_flint.cpp]: There is no function for getting a hash in Flint so I convert to string.

Comment (rwst) [inline symengine/series_flint.cpp]: Ah you mean the extra `c_str()` and seed? Will try without.

Comment (certik) [inline symengine/series_flint.cpp]: Yes, that's what I meant.

File: CMakeLists.txt (hunk 2 above)
Comment (certik) [inline CMakeLists.txt]: I've reworked how this is handled in #757, it works great. So don't worry about it for now, just make sure it works. Once it gets in, I'll then merge it with my PR.

**Comment by certik** [comment]
@rwst how does it compare to your series benchmarks that you have on github? Is there some overhead in SymEngine? If so, we need to fix it.

**Comment by rwst** [comment]
It's worse because of Flint's creation of the two series `sin(x)` and `cos(x)` as explained above, which takes most of the time.

**Comment by certik** [comment]
There is `series_sin` defined here: https://github.com/symengine/symengine/blob/3f0c0947ae09bac9545334a6903d9d72c7043a52/symengine/series.h#L279, and then you redefine it here: https://github.com/symengine/symengine/pull/780/files#diff-16c8b61b394d518f97214d67f2372e7eR65

But you don't use virtual methods, how does it work? Which method will be called? 

Also, is the first template implementation faster than the one in flint?

**Comment by isuruf** [comment]
@certik, that's the beauty of CRTP. Since we get the class `Series` as a template argument, when you call `Series::series_sin`, it will call the static method in `Series` class. If it was not defined in `Series` class as in Piranha, it will call the static method in the parent of `Series` class

**Comment by certik** [comment]
So in this PR, it will call the Flint's implementation, but if we remove `series_sin`, it will call our generic implementation?

**Comment by isuruf** [comment]
@certik, yes.
Additional info: If you remove `series_cosh`, the generic implementation will call Flint implementation's `series_exp` and `series_invert` instead of the generic implementation.

**Comment by certik** [comment]
Ah I see, nice. That's very cool, I like this design.

**Comment by rwst** [comment]
I made the other experiment and tried `sin(x+x^2)*cos(x+x^2), N=1000` and surprise: Flint 1.7s, Pari 12s, Piranha `URatP` is not finished after 5 minutes... so I scaled down to `N=300`: Flint 0.071s, Pari 0.37s, Piranha `URatP` 11.1s. These are worlds apart.

**Comment by isuruf** [comment]
URatP is slower because of clearing pow caches

**Comment by isuruf** [comment]
Solution would be to clear it only if a higher precedence is called. 
CC @bluescarni

**Comment by rwst** [comment]
I will replace `sin(x)*cos(x)` with `sin(x+x^2)*cos(x+x^2)` in the benchmark and make two separate files for flint and piranha. It makes no sense to have both in one.

**Comment by rwst** [comment]
There is still the wider issue that affects all packages, even Pari: that it is faster to substitute if the argument is small, e.g. Pari:
```
? sin(x+x^2)*cos(x+x^2)
...
? ##
  ***   last result computed in 361 ms.
? subst(sin(x)*cos(x),x,x+x^2)
...
? ##
  ***   last result computed in 76 ms.
```
So someone will have to patch Pari and Flint as well with that optimization (first implemented in ring series).

**Comment by bluescarni** [comment]
@rwst @isuruf I will be able to take a look later today. I would like to understand exactly what is going on, I expect flint to be faster but not by that amount.

**Comment by rwst** [comment]
I would prefer to add optimizations and the integration in the `series()` function in another PR.

File: benchmarks/series_expansion_sincos_flint.cpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    //std::cout << *res[N-1] << std::endl;
+    std::cout
+        << std::chrono::duration_cast<std::chrono::milliseconds>(t2-t1).count()
+        << "ms" << std::endl;
+
+    return 0;
+}
+#else
+int main() {}
+#endif
```
Comment (certik) [inline benchmarks/series_expansion_sincos_flint.cpp]: In here I would simply not build the benchmark if Flint is not installed. We can fix this after the PR is in.

File: benchmarks/series_expansion_sincos_piranha.cpp [hunk 4]
```diff
@@ -39,7 +44,5 @@ int main(int argc, char* argv[])
     return 0;
 }
 #else
-
-int main(int, char*[]) {}
-
+int main() {}
 #endif
```
Comment (certik) [inline benchmarks/series_expansion_sincos_piranha.cpp]: The same here.

File: symengine/series_flint.cpp [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+unsigned int str_hash(const char* s, unsigned int seed = 0)
+{
+    unsigned int hash = seed;
+    while (*s)
+    {
+        hash = hash * 101  +  *s++;
+    }
+    return hash;
+}
```
Comment (certik) [inline symengine/series_flint.cpp]: This function is not used, so you can remove it.

**Comment by certik** [comment]
Just remove `str_hash` and it is +1 to merge from me if tests pass.

**Comment by rwst** [comment]
To be continued.

**Comment by fredrik-johansson** [comment]
Right now `sin(x+x^2+O(x^1000))` with Flint is slow, because it doesn't have a basecase algorithm for short input. You can compare with `exp(x+x^2+O(x^1000))`, which does have a basecase and is 700 times faster. Adding a similar basecase for sin/cos in on the todo list...

**Comment by fredrik-johansson** [comment]
I've fixed this problem in flint -- `sin(x+x^2+O(x^1000))` takes 0.003 seconds now.

Some other functions still need to be optimized this way.

**Comment by certik** [comment]
Thanks Fredrik! Which Flint commits do that?

**Comment by fredrik-johansson** [comment]
The latest commit 57176eb1c02632340076a5f4ad6a69911824be5c in fredrik-johansson/flint2.git. I expect that Bill Hart will merge it in a moment.

**Comment by certik** [comment]
Here is the link: https://github.com/fredrik-johansson/flint2/commit/57176eb1c02632340076a5f4ad6a69911824be5c

**Comment by fredrik-johansson** [comment]
Thanks. It's in wbhart/flint2 now.

**Comment by certik** [comment]
@fredrik-johansson according to your opinion now, if we expand sin or cos using Flint, does it now run at optimal speed? Or is there still space for improvement?

**Comment by fredrik-johansson** [comment]
There are still some improvements that can be done in Flint, but we're talking about very small constant factor overhead (< 2).

**Comment by certik** [comment]
Ok, perfect.

**Comment by fredrik-johansson** [comment]
Actually, it's not going to be optimal for certain very sparse inputs, like
x + x^100 or indeed even x^100. You can always add more special cases for
such things. I think this is of somewhat limited practical utility, though.

**Comment by certik** [comment]
Ok. How about multivariate polynomials? We've seen huge speedups in Piranha compared to Flint for multivariate polynomials, especially sparse. For single variate ones, Flint seems to be faster, as Piranha is not optimized for those.

**Comment by fredrik-johansson** [comment]
There's no code in Flint for multivariate polynomials. There have been some
plans, but no one to work on them.

**Comment by certik** [comment]
Nemo can do them.

**Comment by fredrik-johansson** [comment]
Nemo has (dense) univariate polynomials, which allows you to create
polynomials of polynomials. This is fine for representing dense
multivariate polynomials in a small number of variables.

Fredrik

**Pull request merged**